### PR TITLE
Change AdamW step to a tensor instead of an int

### PR DESCRIPTION
**Pull request by eracah**
# What does this PR do?
changes the step in AdamW to a tensor

# What issue(s) does this change relate to?
this avoids this issue: https://github.com/pytorch/pytorch/issues/99079 and this one: https://github.com/mosaicml/composer/issues/2231 because by setting step to a tensor, there are no more non-tensors in the optimizer state, so we can't get to the buggy part of the torch code.

fix CO-2107

### CODE REVIEW

**Review by dakinggg — COMMENTED** [review]
Can you add a test that would've caught this? I guess fsdp checkpoint with Adam?

**Comment by eracah** [comment]
> Can you add a test that would've caught this? I guess fsdp checkpoint with Adam?

Ok added it. didn't run it, so will let the CI determine if it passes

File: tests/trainer/test_sharded_checkpoint.py [hunk 1]
```diff
@@ -81,8 +86,11 @@ def get_trainer(
 
 def _compare_optims_between_state_dicts(state_dict1, state_dict2):
     # Check that optim params are equal between checkpoint and in memory optimizer
-    state_dict1_optim_params = state_dict1['optimizers']['Adam']['state']
-    state_dict2_optim_params = state_dict2['optimizers']['Adam']['state']
+    optim_key1 = list(state_dict1['optimizers'].keys()).pop()
```
Comment (dakinggg) [inline tests/trainer/test_sharded_checkpoint.py:89]: maybe assert these lists are of length 1, just to make the assumption more clear

**Review by dakinggg — APPROVED** [review]

File: tests/trainer/test_sharded_checkpoint.py (hunk 1 above)
Comment (eracah) [reply tests/trainer/test_sharded_checkpoint.py:89]: haha we've lost years of our lives checking if the number of optimizers is 1

**Comment by shivashankarrs** [comment]
> Can you add a test that would've caught this? I guess fsdp checkpoint with Adam?

Thanks Daniel :) Hi here again.

**Comment by dakinggg** [comment]
Hey @shivashankarrs :)

**Pull request merged**